File: src/main/java/com/morph/engine/math/Vector2f.java
package com.morph.engine.math;

import java.util.function.Function;

public class Vector2f {
	private float x, y;
	public Vector2f(float x, float y) {
		this.x = x;
		this.y = y;
	}
	
	/**
	 * Clones a Vector2f.
	 */
	public Vector2f(Vector2f v) {
		this.x = v.getX();
		this.y = v.getY();
	}
	
	public Vector2f() {
		this.x = 0;
		this.y = 0;
	}
	
	public Vector2f(double x, double y) {
		this.x = (float)x;
		this.y = (float)y;
	}

	public String toString() {
		return "Vector2f(" + x + ", " + y + ")";
	}
	
	@Override
	public boolean equals(Object o) {
		if (o instanceof Vector2f) {
			Vector2f v = (Vector2f) o;
			float epsilon = 1e-8f;
			
			return Math.abs(x - v.getX()) < epsilon && Math.abs(y - v.getY()) < epsilon;
		}
		
		return false;
	}
	
	public Vector2f add(Vector2f v) {
		return new Vector2f(x + v.getX(), y + v.getY());
	}
	
	public Vector2f sub(Vector2f v) {
		return new Vector2f(x - v.getX(), y - v.getY());
	}
	
	public Vector2f mul(Vector2f v) {
		return new Vector2f(x * v.getX(), y * v.getY());
	}
	
	public Vector2f div(Vector2f v) {
		return new Vector2f(x / v.getX(), y / v.getY());
	}
	
	public Vector2f scale(float k) {
		return new Vector2f(x * k, y * k);
	}
	
	public Vector2f invScale(float k) {
		return new Vector2f(x / k, y / k);
	}
	
	public float getLength() {
		return (float) Math.sqrt(x*x + y*y);
	}
	
	public Vector2f normalize() {
		float length = getLength();
		return this.invScale(length);
	}
	
	public Vector2f abs() {
		return new Vector2f(Math.abs(x), Math.abs(y));
	}

	public float getX() {
		return x;
	}
	
	public float getY() {
		return y;
	}
	
	public void setX(float x) {
		this.x = x;
	}
	
	public void setY(float y) {
		this.y = y;
	}
	
	public void set(float x, float y) {
		this.x = x;
		this.y = y;
	}
	
	public void set(Vector2f v) {
		this.x = v.getX();
		this.y = v.getY();
	}

	public Matrix4f asTranslationMatrix() {
		return new Matrix4f(
				1, 0, 0, x,
				0, 1, 0, y,
				0, 0, 1, 0,
				0, 0, 0, 1
		);
	}
	
	public Matrix4f asScaleMatrix() {
		return new Matrix4f(
				x, 0, 0, 0,
				0, y, 0, 0,
				0, 0, 1, 0,
				0, 0, 0, 1
		);
	}

	public float cross(Vector2f v) {
		return x * v.getY() - y * v.getX();
	}

	public void clear() {
		this.x = 0;
		this.y = 0;
	}

	public Vector2f negate() {
		return new Vector2f(-x, -y);
	}

	public float dot(Vector2f v) {
		return (x * v.getX()) + (y * v.getY());
	}

	public Vector2f map(Function<Float, Float> fun) {
		return new Vector2f(fun.apply(x), fun.apply(y));
	}

	public static Vector2f reflect(Vector2f n, Vector2f v) {
		return v.add(n.scale(-2 * v.dot(n)));
	}
}

File: src/main/kotlin/com/morph/engine/math/Vector2f.kt
package com.morph.engine.math

data class Vector2f(var x: Float = 0f, var y: Float = 0f) {
    val length: Float
        get() = Math.sqrt((x * x + y * y).toDouble()).toFloat()

    /**
     * Clones a Vector2f.
     */
    constructor(v: Vector2f) : this(v.x, v.y)

    override fun toString(): String {
        return "Vector2f($x, $y)"
    }

    @JvmName("add")
    operator fun plus(v: Vector2f): Vector2f {
        return Vector2f(x + v.x, y + v.y)
    }

    @JvmName("sub")
    operator fun minus(v: Vector2f): Vector2f {
        return Vector2f(x - v.x, y - v.y)
    }

    @JvmName("mul")
    operator fun times(v: Vector2f): Vector2f {
        return Vector2f(x * v.x, y * v.y)
    }

    operator fun div(v: Vector2f): Vector2f {
        return Vector2f(x / v.x, y / v.y)
    }

    @JvmName("scale")
    operator fun times(k: Float): Vector2f {
        return Vector2f(x * k, y * k)
    }

    @JvmName("invScale")
    operator fun div(k: Float): Vector2f {
        return Vector2f(x / k, y / k)
    }

    fun normalize(): Vector2f {
        val length = length
        return this / length
    }

    fun abs(): Vector2f {
        return Vector2f(Math.abs(x), Math.abs(y))
    }

    operator fun set(x: Float, y: Float) {
        this.x = x
        this.y = y
    }

    fun set(v: Vector2f) {
        this.x = v.x
        this.y = v.y
    }

    fun asTranslationMatrix(): Matrix4f {
        return Matrix4f(
                1f, 0f, 0f, x,
                0f, 1f, 0f, y,
                0f, 0f, 1f, 0f,
                0f, 0f, 0f, 1f
        )
    }

    fun asScaleMatrix(): Matrix4f {
        return Matrix4f(
                x, 0f, 0f, 0f,
                0f, y, 0f, 0f,
                0f, 0f, 1f, 0f,
                0f, 0f, 0f, 1f
        )
    }

    fun cross(v: Vector2f): Float {
        return x * v.y - y * v.x
    }

    fun clear() {
        this.x = 0f
        this.y = 0f
    }

    @JvmName("negate")
    operator fun unaryMinus(): Vector2f {
        return Vector2f(-x, -y)
    }

    infix fun dot(v: Vector2f): Float {
        return x * v.x + y * v.y
    }

    fun map(f: (Float) -> Float) = Vector2f(f(x), f(y))

    companion object {
        fun reflect(n: Vector2f, v: Vector2f): Vector2f {
            return v + (n * (-2 * (v dot n)))
        }
    }
}
